File: src/backend/trip-service/app/api/v1/trips.py
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from app.core.config import get_settings
from app.services import trip_service, maps_service, dispatch_service

router = APIRouter(prefix="/trips", tags=["trips"])


class CreateTripRequest(BaseModel):
    driver_phone:    str
    customer_name:   str
    customer_phone:  str = ""
    pickup_address:  str
    dropoff_address: str
    pickup_time:     str = ""
    booked_via:      str = "app"   # app | voice | admin


class UpdateStatusRequest(BaseModel):
    status: str   # confirmed | rejected | no_answer | completed


# ── Create trip ───────────────────────────────────────────────────────────────

@router.post("", status_code=201)
async def create_trip(body: CreateTripRequest):
    s = get_settings()

    route = await maps_service.get_route(
        body.pickup_address,
        body.dropoff_address,
        s.google_maps_api_key,
    )

    trip = trip_service.create({
        **body.model_dump(),
        "distance_km":   route.get("distance_km", 0),
        "eta_minutes":   route.get("eta_minutes", 0),
        "traffic_note":  route.get("traffic_note", ""),
        "route_summary": route.get("summary", ""),
    })
    return trip


# ── List trips (admin + polling) ─────────────────────────────────────────────

@router.get("")
def list_trips(
    status:         str = Query(default=""),
    driver_phone:   str = Query(default=""),
    customer_phone: str = Query(default=""),
    limit:          int = Query(default=50, le=200),
):
    """
    GET /api/v1/trips                                → all trips (admin)
    GET /api/v1/trips?status=pending                 → filter by status
    GET /api/v1/trips?driver_phone=+84xxx            → driver's trips
    GET /api/v1/trips?customer_phone=+84xxx          → customer's trips (for AI agent)
    """
    if driver_phone and status:
        return trip_service.list_by_driver_and_status(driver_phone, status, limit)
    if driver_phone:
        return trip_service.list_by_driver(driver_phone, limit)
    if customer_phone:
        return trip_service.list_by_customer(customer_phone, limit)
    if status:
        return trip_service.list_by_status(status, limit)
    return trip_service.list_recent(limit)


# ── Get one trip ──────────────────────────────────────────────────────────────

@router.get("/{trip_id}")
def get_trip(trip_id: str):
    trip = trip_service.get(trip_id)
    if not trip:
        raise HTTPException(status_code=404, detail="Trip not found")
    return trip


# ── Update status ─────────────────────────────────────────────────────────────

@router.patch("/{trip_id}")
def update_trip_status(trip_id: str, body: UpdateStatusRequest):
    allowed = {"confirmed", "rejected", "no_answer", "calling", "notified", "pending", "completed"}
    if body.status not in allowed:
        raise HTTPException(status_code=400, detail=f"status must be one of {allowed}")

    trip = trip_service.get(trip_id)
    if not trip:
        raise HTTPException(status_code=404, detail="Trip not found")

    return trip_service.update_status(trip_id, body.status)


# ── Dispatch call (fallback) ──────────────────────────────────────────────────

@router.post("/{trip_id}/dispatch")
async def dispatch_trip(trip_id: str):
    """Trigger outbound SIP call to driver — used as fallback when driver has no app."""
    trip = trip_service.get(trip_id)
    if not trip:
        raise HTTPException(status_code=404, detail="Trip not found")
    if trip["status"] not in ("pending", "notified", "rejected", "no_answer"):
        raise HTTPException(status_code=409, detail=f"Cannot dispatch trip with status={trip['status']}")

    try:
        room_name = await dispatch_service.dispatch_call(dict(trip))
    except Exception as exc:
        raise HTTPException(status_code=500, detail=str(exc))

    return trip_service.update_status(trip_id, "calling", {"room_name": room_name})


# ── Notify driver (app push / polling) ───────────────────────────────────────

@router.post("/{trip_id}/notify")
async def notify_driver(trip_id: str):
    """
    Mark trip as notified (driver app will pick it up via polling).
    In production: also sends FCM push here.
    """
    trip = trip_service.get(trip_id)
    if not trip:
        raise HTTPException(status_code=404, detail="Trip not found")

    return trip_service.update_status(trip_id, "notified")

File: src/backend/trip-service/src/routes/trips.ts
import { Router, Request, Response } from 'express';
import * as tripService from '../services/tripService';
import * as mapsService from '../services/mapsService';
import * as dispatchService from '../services/dispatchService';
import { config } from '../config';
import { TripData } from '../services/dispatchService';

const router = Router();

const ALLOWED_STATUSES = new Set(['confirmed', 'rejected', 'no_answer', 'calling', 'notified', 'pending', 'completed']);

router.post('/', async (req: Request, res: Response) => {
  try {
    const body = req.body;
    const route = await mapsService.getRoute(body.pickup_address, body.dropoff_address, config.googleMapsApiKey);
    const trip = await tripService.create({
      ...body,
      distance_km: route.distance_km ?? 0,
      eta_minutes: route.eta_minutes ?? 0,
      traffic_note: route.traffic_note ?? '',
      route_summary: route.summary ?? '',
    });
    return res.status(201).json(trip);
  } catch (err) {
    return res.status(500).json({ error: String(err) });
  }
});

router.get('/', async (req: Request, res: Response) => {
  try {
    const limit = Math.min(Number(req.query.limit) || 50, 200);
    const status = req.query.status ? String(req.query.status) : '';
    const driverPhone = req.query.driver_phone ? String(req.query.driver_phone) : '';
    const customerPhone = req.query.customer_phone ? String(req.query.customer_phone) : '';

    if (driverPhone && status) return res.json(await tripService.listByDriverAndStatus(driverPhone, status, limit));
    if (driverPhone) return res.json(await tripService.listByDriver(driverPhone, limit));
    if (customerPhone) return res.json(await tripService.listByCustomer(customerPhone, limit));
    if (status) return res.json(await tripService.listByStatus(status, limit));
    return res.json(await tripService.listRecent(limit));
  } catch (err) {
    return res.status(500).json({ error: String(err) });
  }
});

router.get('/:tripId', async (req: Request, res: Response) => {
  try {
    const trip = await tripService.get(req.params.tripId);
    if (!trip) return res.status(404).json({ detail: 'Trip not found' });
    return res.json(trip);
  } catch (err) {
    return res.status(500).json({ error: String(err) });
  }
});

router.patch('/:tripId', async (req: Request, res: Response) => {
  try {
    const { status } = req.body;
    if (!ALLOWED_STATUSES.has(status)) {
      return res.status(400).json({ detail: `status must be one of ${[...ALLOWED_STATUSES].join(', ')}` });
    }
    const trip = await tripService.get(req.params.tripId);
    if (!trip) return res.status(404).json({ detail: 'Trip not found' });
    return res.json(await tripService.updateStatus(req.params.tripId, status));
  } catch (err) {
    return res.status(500).json({ error: String(err) });
  }
});

router.post('/:tripId/dispatch', async (req: Request, res: Response) => {
  try {
    const trip = await tripService.get(req.params.tripId);
    if (!trip) return res.status(404).json({ detail: 'Trip not found' });
    if (!['pending', 'notified', 'rejected', 'no_answer'].includes(trip.status)) {
      return res.status(409).json({ detail: `Cannot dispatch trip with status=${trip.status}` });
    }
    const roomName = await dispatchService.dispatchCall(trip as TripData);
    return res.json(await tripService.updateStatus(req.params.tripId, 'calling', { room_name: roomName }));
  } catch (err) {
    return res.status(500).json({ detail: String(err) });
  }
});

router.post('/:tripId/notify', async (req: Request, res: Response) => {
  try {
    const trip = await tripService.get(req.params.tripId);
    if (!trip) return res.status(404).json({ detail: 'Trip not found' });
    return res.json(await tripService.updateStatus(req.params.tripId, 'notified'));
  } catch (err) {
    return res.status(500).json({ error: String(err) });
  }
});

export default router;
